sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -36,75 +38,345 @@ class GpuGenerateExecSparkPlanMeta(
- * where we don't actually need to put the data into an array first.
+ * GPU overrides of `Generator`, corporate with `GpuGenerateExec`.
  */
+trait GpuGenerator extends GpuUnevaluable {
+
+  override def dataType: DataType = ArrayType(elementSchema)
+
+  override def foldable: Boolean = false
+
+  override def nullable: Boolean = false
+
+  /**
+   * The output element schema.
+   */
+  def elementSchema: StructType
+
+  /**
+   * Apply generator to produce result ColumnarBatch from input batch.
+   *
+   * This is a specialized method for GPU runtime, which is called by GpuGenerateExec who owns
+   * the generator. The reason of creating a new method rather than implementing columnarEval is
+   * that generator is an integrated Table transformer instead of column transformer in terms of
+   * cuDF.
+   *
+   * @param inputBatch      projected input data, which ensures appending columns are ahead of
+   *                        generators' inputs. So, generators can distinguish them with an offset.
+   * @param generatorOffset column offset of generator's input columns in `inputBatch`
+   * @param outer           when true, each input row will be output at least once, even if the
+   *                        output of the given `generator` is empty.
+   * @return result ColumnarBatch
+   */
+  def generate(inputBatch: ColumnarBatch,
+    generatorOffset: Int,
+    outer: Boolean): ColumnarBatch
+
+  /**
+   * Compute split indices for generator's input batches.
+   *
+   * This is a specialized method for GPU runtime, which is called by GpuGenerateExec to split up
+   * input batches to reduce total memory cost during generating. It is necessary because most of
+   * generators may produce multiple records for each input record, which make output batch size
+   * much larger than input size.
+   *
+   * @param inputBatch      projected input data, which ensures appending columns are ahead of
+   *                        generators' inputs. So, generators can distinguish them with an offset.
+   * @param generatorOffset column offset of generator's input columns in `inputBatch`
+   * @param outer           when true, each input row will be output at least once, even if the
+   *                        output of the given `generator` is empty.
+   * @param targetSizeBytes the target number of bytes for a GPU batch, one of `RapidsConf`
+   * @return split indices of input batch
+   */
+  def inputSplitIndices(inputBatch: ColumnarBatch,
+      generatorOffset: Int,
+      outer: Boolean,
+      targetSizeBytes: Long): Array[Int]
+
+  /**
+   * Extract lazy expressions from generator if exists.
+   *
+   * This is a specialized method for GPU runtime, which is called by GpuGenerateExec to determine
+   * whether current generation plan can be executed with optimized lazy array generation or not.
+   *
+   * @return fixed length lazy expressions for generation
+   */
+  def fixedLenLazyExpressions: Seq[Expression] = Nil
```

Comment:
nit: we should document that Nil means there are no expressions to extract.

Reply:
Done.